### ISSUE TITLE
SIMD reduction pragmas throw error with Intel Compiler

### ISSUE THREAD

**Issue by fcdimitr** [zhang416/recfmm#1]
I am attaching the error that is generated by the icc compiler

```
fmm-action.c(2655): error: "simd" region may not be closely nested inside a "simd" region
      #pragma simd reduction(+:pot)
              ^

fmm-action.c(2656): error: "simd" region may not be closely nested inside a "simd" region
      #pragma simd reduction(+:fx)
              ^

fmm-action.c(2657): error: "simd" region may not be closely nested inside a "simd" region
      #pragma simd reduction(+:fy)
```

The corresponding lines of code are the following ones (for Laplace and Yukawa kernels, respectively).

```
@@ -2652,10 +2652,10 @@ void direct_evaluation(const fmm_box_t *tbox, const fmm_box_t *sbox) {
     double pot = 0, fx = 0, fy = 0, fz = 0; 
 
 #ifdef icc
    #pragma simd reduction(+:pot)
    #pragma simd reduction(+:fx)
    #pragma simd reduction(+:fy)
    #pragma simd reduction(+:fz)
 #endif
     for (int j = start2; j <= end2; j++) {
       int j3 = j * 3; 
@@ -4876,10 +4876,10 @@ void direct_evaluation(const fmm_box_t *tbox, const fmm_box_t *sbox) {
     double pot = 0, fx = 0, fy = 0, fz = 0; 
 
 #ifdef icc
    #pragma simd reduction(+:pot)
    #pragma simd reduction(+:fx)
    #pragma simd reduction(+:fy)
    #pragma simd reduction(+:fz)
 #endif
     for (int j = start2; j <= end2; j++) {
       int j3 = j * 3; 

```
I have verified that the compilation succeeds and the results are correct if the simd pragmas are removed. 

The issue was encountered using `icpc version 18.0.2 (gcc version 7.0.0 compatibility)`.  Compilation with `gcc version 7.3.0` is successful (since those pragmas are only used for the Intel compiler).

**Comment by zhang416** [comment]
For old versions of intel compiler such as 15.0.3, the code still compiles. The error is probably due to this:  https://www.openmp.org/about/openmp-faq/, search "closely-nested". 

```
zhang416@login1:~/recfmm> icc --version
icc (ICC) 15.0.3 20150407
Copyright (C) 1985-2015 Intel Corporation.  All rights reserved.
zhang416@login1:~/recfmm> which icc
/N/soft/cle5/intel/composer_xe_2015.3.187/bin/intel64/icc
zhang416@login1:~/recfmm> make CC=icc TBBLOC=/N/soft/cle5/intel/composer_xe_2015.3.187/bin/intel64/tbb
make clean laplace
make[1]: Entering directory `/gpfs/home/z/h/zhang416/BigRed2/recfmm'
(cd ./src; make clean)
make[2]: Entering directory `/gpfs/home/z/h/zhang416/BigRed2/recfmm/src'
rm -f *.o *~ libadap_laplace.a libadap_yukawa.a
make[2]: Leaving directory `/gpfs/home/z/h/zhang416/BigRed2/recfmm/src'
(cd ./example; make clean)
make[2]: Entering directory `/gpfs/home/z/h/zhang416/BigRed2/recfmm/example'
rm -f *.o example *~
make[2]: Leaving directory `/gpfs/home/z/h/zhang416/BigRed2/recfmm/example'
(cd ./test; make clean)
make[2]: Entering directory `/gpfs/home/z/h/zhang416/BigRed2/recfmm/test'
rm -f fmm *.o *~
make[2]: Leaving directory `/gpfs/home/z/h/zhang416/BigRed2/recfmm/test'
(cd ./lib; rm -f *.a)
/bin/sh: line 0: cd: ./lib: No such file or directory
rm -rf *~ ./lib
cd ./src; make lib_lap
make[2]: Entering directory `/gpfs/home/z/h/zhang416/BigRed2/recfmm/src'
icc -pedantic -O3 -std=gnu99 -DUSING_CILK -Dicc -I../include -I/N/soft/cle5/intel/composer_xe_2015.3.187/bin/intel64/tbb/include -DLAPLACE -c fmm-action.c fmm-dag.c fmm-param.c
ar -cvq libadap_laplace.a fmm-action.o fmm-dag.o fmm-param.o
a - fmm-action.o
a - fmm-dag.o
a - fmm-param.o
make[2]: Leaving directory `/gpfs/home/z/h/zhang416/BigRed2/recfmm/src'
mkdir -p lib
cp ./src/libadap_laplace.a ./lib/
cd ./test; make clean lap laptest3 laptest6
make[2]: Entering directory `/gpfs/home/z/h/zhang416/BigRed2/recfmm/test'
rm -f fmm *.o *~
icc -Wall -pedantic -O3 -std=gnu99 -DUSING_CILK -Dicc -I../include -I/N/soft/cle5/intel/composer_xe_2015.3.187/bin/intel64/tbb/include   -DLAPLACE -c fmm-test.c -o fmm-test.o 
icc -o fmm fmm-test.o -L/N/soft/cle5/intel/composer_xe_2015.3.187/bin/intel64/tbb/lib -ltbbmalloc -lcilkrts -lm -L../lib -ladap_laplace
./fmm -d 1 -a 3
test configuration:
  number of sources: 10000
  number of targets: 10000
  accuracy requirement: 40
accuracy statistics:
  error of potential in L-2 norm: 3.118752e-04
  error of potential in L-infty norm: 7.475367e-02
  error of field in L-2 norm: 3.623090e-04
******* pass *******
./fmm -d 2 -a 3
test configuration:
  number of sources: 10000
  number of targets: 10000
  accuracy requirement: 40
accuracy statistics:
  error of potential in L-2 norm: 8.322567e-05
  error of potential in L-infty norm: 1.585691e-01
  error of field in L-2 norm: 9.463150e-06
******* pass *******
./fmm -d 1 -a 6
test configuration:
  number of sources: 10000
  number of targets: 10000
  accuracy requirement: 40
accuracy statistics:
  error of potential in L-2 norm: 1.772826e-07
  error of potential in L-infty norm: 4.208684e-05
  error of field in L-2 norm: 4.719721e-07
******* pass *******
./fmm -d 2 -a 6
test configuration:
  number of sources: 10000
  number of targets: 10000
  accuracy requirement: 40
accuracy statistics:
  error of potential in L-2 norm: 5.796298e-08
  error of potential in L-infty norm: 8.826200e-05
  error of field in L-2 norm: 6.512689e-09
******* pass *******
make[2]: Leaving directory `/gpfs/home/z/h/zhang416/BigRed2/recfmm/test'
make[1]: Leaving directory `/gpfs/home/z/h/zhang416/BigRed2/recfmm'
make clean yukawa
make[1]: Entering directory `/gpfs/home/z/h/zhang416/BigRed2/recfmm'
(cd ./src; make clean)
make[2]: Entering directory `/gpfs/home/z/h/zhang416/BigRed2/recfmm/src'
rm -f *.o *~ libadap_laplace.a libadap_yukawa.a
make[2]: Leaving directory `/gpfs/home/z/h/zhang416/BigRed2/recfmm/src'
(cd ./example; make clean)
make[2]: Entering directory `/gpfs/home/z/h/zhang416/BigRed2/recfmm/example'
rm -f *.o example *~
make[2]: Leaving directory `/gpfs/home/z/h/zhang416/BigRed2/recfmm/example'
(cd ./test; make clean)
make[2]: Entering directory `/gpfs/home/z/h/zhang416/BigRed2/recfmm/test'
rm -f fmm *.o *~
make[2]: Leaving directory `/gpfs/home/z/h/zhang416/BigRed2/recfmm/test'
(cd ./lib; rm -f *.a)
rm -rf *~ ./lib
cd ./src; make lib_yuk
make[2]: Entering directory `/gpfs/home/z/h/zhang416/BigRed2/recfmm/src'
icc -pedantic -O3 -std=gnu99 -DUSING_CILK -Dicc -I../include -I/N/soft/cle5/intel/composer_xe_2015.3.187/bin/intel64/tbb/include -DYUKAWA -c fmm-action.c fmm-dag.c fmm-param.c
ifort -O3 -c ribesl.f dgamma.f
ar -cvq libadap_yukawa.a fmm-action.o fmm-dag.o fmm-param.o ribesl.o dgamma.o
a - fmm-action.o
a - fmm-dag.o
a - fmm-param.o
a - ribesl.o
a - dgamma.o
make[2]: Leaving directory `/gpfs/home/z/h/zhang416/BigRed2/recfmm/src'
mkdir -p lib
cp ./src/libadap_yukawa.a ./lib/
cd ./test; make clean yuk yuktest3 yuktest6
make[2]: Entering directory `/gpfs/home/z/h/zhang416/BigRed2/recfmm/test'
rm -f fmm *.o *~
icc -Wall -pedantic -O3 -std=gnu99 -DUSING_CILK -Dicc -I../include -I/N/soft/cle5/intel/composer_xe_2015.3.187/bin/intel64/tbb/include   -DYUKAWA -c fmm-test.c -o fmm-test.o
ifort  -nofor_main -o fmm fmm-test.o -L/N/soft/cle5/intel/composer_xe_2015.3.187/bin/intel64/tbb/lib -ltbbmalloc -lcilkrts -lm -L../lib -ladap_yukawa	
./fmm -d 1 -a 3
test configuration:
  number of sources: 10000
  number of targets: 10000
  accuracy requirement: 40
accuracy statistics:
  error of potential in L-2 norm: 3.105479e-04
  error of potential in L-infty norm: 1.155696e+00
  error of field in L-2 norm: 3.575760e-04
******* pass *******
./fmm -d 2 -a 3
test configuration:
  number of sources: 10000
  number of targets: 10000
  accuracy requirement: 40
accuracy statistics:
  error of potential in L-2 norm: 8.194450e-05
  error of potential in L-infty norm: 2.469407e+00
  error of field in L-2 norm: 9.445127e-06
******* pass *******
./fmm -d 1 -a 6
test configuration:
  number of sources: 10000
  number of targets: 10000
  accuracy requirement: 40
accuracy statistics:
  error of potential in L-2 norm: 1.771219e-07
  error of potential in L-infty norm: 6.330990e-04
  error of field in L-2 norm: 5.376775e-07
******* pass *******
./fmm -d 2 -a 6
test configuration:
  number of sources: 10000
  number of targets: 10000
  accuracy requirement: 40
accuracy statistics:
  error of potential in L-2 norm: 5.598763e-08
  error of potential in L-infty norm: 1.389032e-03
  error of field in L-2 norm: 6.645550e-09
******* pass *******
make[2]: Leaving directory `/gpfs/home/z/h/zhang416/BigRed2/recfmm/test'
make[1]: Leaving directory `/gpfs/home/z/h/zhang416/BigRed2/recfmm'
```

**Comment by zhang416** [comment]
Addressed in #3
